Create eip-5018 (#5018)
ethereum/EIPs: EIPS/eip-5018.md

---
eip: 5018
title: Directory Standard
description: A standard interface for filesystem directories.
author: Qi Zhou (@qizhou)
discussions-to: TBD
status: Draft
type: Standards Track
category: ERC
created: 2022-04-18
---

## Abstract

The following standard allows for the implementation of a standard API for filesystem directories within smart contracts.
This standard provides basic functionality to read/write binary objects for any size, as well as allow reading/writing chunks of the object if the object is too large to fit in a single transaction.

## Motivation

> A standard interface allows any binary objects on EVM-based blockchain to be re-used by other dApps.

Reviewer: Would you mind going a bit more in depth on why this EIP is necessary? Is storing a tree of binary data on-chain really that common?

Author: Thanks for the comment.  I have added the motivation with the applications with EIP-4804.  With the standard, we can put web contents (e.g., HTML/SVG) in the same tree structure as the local directory and reference them using protocol-relative URLs, i.e, web3:// defined by EIP-4804.

Please let me know if the motivation is clear or not.

Author: Just uploaded the homepages of ENS and Uniswap on Rinkeby using the directory standard and EIP-4804:

- ENS:
https://galileo.web3q.io/ensdomains.eth:4/
or 
https://galileo.web3q.io/0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b:4/

https://galileo.web3q.io/uniswaps.eth:4/
or
https://galileo.web3q.io/0xC100d49e8F3d621E4438E82E4f95CF505b3E2a28:4/

> With EIP-4804, we are able to locate a Web3 resource on blockchain using HTTP-style URIs.  One application of Web3 resources are web contents that are referenced within a directory using relative paths such as HTML/SVG.  This standard proposes a contract-based directory to simplify the mapping between local web contents and on-chain web contents.  Further, with relative paths referenced in the web contents and EIP-4804, the users will have a consistent view of the web contents locally and on-chain.

Reviewer: The first mention of another EIP should link to it.

```suggestion
With [EIP-4804](./eip-4804.md), we are able to locate a Web3 resource on blockchain using HTTP-style URIs.  One application of Web3 resources are web contents that are referenced within a directory using relative paths such as HTML/SVG.  This standard proposes a contract-based directory to simplify the mapping between local web contents and on-chain web contents.  Further, with relative paths referenced in the web contents and EIP-4804, the users will have a consistent view of the web contents locally and on-chain.
```

## Specification

### Directory

> ### Methods

Reviewer: Is this supposed to be one level deeper?

```suggestion
#### Methods
```

```
function write(bytes memory name, bytes memory data) external payable
```

#### read

Returns the binary `data` from the file `name` in the directory and existence of the file.

```
function read(bytes memory name) external view returns (bytes memory data, bool exist)
```

#### fallback read

Returns the binary `data` from the file `prefixedName` (prefixed with `/`) in the directory.

```
fallback(bytes calldata prefixedName) external returns (bytes memory data) 
```

#### size

Returns the size of the `data` from the file `name` in the directory and the number of chunks of the data.

```
function size(bytes memory name) external view returns (uint256 size, uint256 chunks)
```

#### remove

Removes the file `name` in the directory and returns the number of chunks removed (0 means the file does not exist) by an account with write permission.

```
function remove(bytes memory name) external returns (uint256 numOfChunksRemoved)
```

#### countChunks

Returns the number of chunks of the file `name`.

```
function countChunks(bytes memory name) external view returns (uint256 numOfChunks);
```

#### writeChunk

> Writes a chunk of data to the file by an account with write permission. The write will fail if `chunkId > numOfChunks`, i.e., the write must append the file or replace the existing chunk.

Reviewer: How does this interact with the unchunked functions (`read`/`write`)? For example, if I `writeChunk` a file, then `read` it, what happens?

Author: Thanks for the question.  `read` should return the concatenated chunked data written by `writeChunk`.  The following gives some examples

- `read("hello.txt")` => "" (file is empty)
- `writeChunk("hello.txt", 0, "abc")`  will succeed
- `read("hello.txt")` => "abc"
- `writeChunk("hello.txt", 1, "efg")`  will succeed
- `read("hello.txt")` => "abcefg"
- `writeChunk("hello.txt", 0, "aaa")`  will succeed (replace chunk 0's data)
- `read("hello.txt")` => "aaaefg"
- `writeChunk("hello.txt", 3, "hij")`  will fail because the operation is not replacement or append.

With `writeChunk` method, we allow writing a file with external data that exceeds the current calldata limit (e.g., 1.8MB now, and 1.4MB post-EIP-4488), and it is able to read the whole file in a single `read` method (which is friendly for large web objects such as HTML/SVG/PNG/JPG, etc).

For `write` method, calling a `write` method will replace all data chunks of the file with `write` method data, and one implementation can be: 1. `writeChunk(filename, chunkId=0, data_from_write)` to chunk 0 with the same `write` method data, and `truncate(filename, chunkId=1)`, which will remove the rest chunks.

Hope this could clarify your question.  Free to let me know if some updates are needed to clarify this.

Reviewer: Great information. I think it deserves to be in the EIP.

Author: Thanks for the suggestion.  I have added a subsection in Rationale with this part.  Please let me know if you have further questions :)

```
 function writeChunk(bytes memory name, uint256 chunkId, bytes memory chunkData) external payable;
```

#### readChunk

Returns the chunk data of the file `name` and the existence of the chunk.

```
function readChunk(bytes memory name, uint256 chunkId) external view returns (bytes memory chunkData, bool exist);
```

#### chunkSize

Returns the size of a chunk of the file `name` and the existence of the chunk.

```
function chunkSize(bytes memory name, uint256 chunkId) external view returns (uint256 chunkSize, bool exist);
```

#### removeChunk

Removes a chunk of the file `name` and returns `false` if such chunk does not exist. The method should be called by an account with write permission.

```
function removeChunk(bytes memory name, uint256 chunkId) external returns (bool exist);
```

#### truncate

Removes the chunks of the file `name` in the directory from the given `chunkId` and returns the number of chunks removed by an account with write permission.  When `chunkId = 0`, the method is essentially the same as `remove()`.

```
function truncate(bytes memory name, uint256 chunkId) external returns (uint256 numOfChunksRemoved);
```

> ## Rationle

Reviewer: ```suggestion
## Rationale
```

One issue of uploading the web contents to the blockchain is that the web contents may be too large to fit into a single transaction.  As a result, the standard provides a chunk-based operations so that uploading a content can be splitted into several transactions.  Meanwhile, the read operation can be done in a single transaction, i.e., with a single Web3 URL defined in EIP-4804.

> 

Reviewer: ```suggestion

## Security Considerations
```

This section is not optional.

---
eip: 5018
title: Directory Standard
description: A standard interface for filesystem directories.
author: Qi Zhou (@qizhou)
discussions-to: https://ethereum-magicians.org/t/eip-5018-directory-standard/8958
status: Draft
type: Standards Track
category: ERC
created: 2022-04-18
---

## Abstract

> The following standard allows for the implementation of a standard API for filesystem directories within smart contracts.
> This standard provides basic functionality to read/write binary objects for any size, as well as allow reading/writing chunks of the object if the object is too large to fit in a single transaction.

Reviewer: ```suggestion
The following standardizes an API for directories and files within smart contracts, similar to traditional filesystems.
This standard provides basic functionality to read/write binary objects of any size, as well as allow reading/writing chunks of the object if the object is too large to fit in a single transaction.
```

```
function removeChunk(bytes memory name, uint256 chunkId) external returns (bool exist);
```

#### truncate

Removes the chunks of the file `name` in the directory from the given `chunkId` and returns the number of chunks removed by an account with write permission.  When `chunkId = 0`, the method is essentially the same as `remove()`.

```
function truncate(bytes memory name, uint256 chunkId) external returns (uint256 numOfChunksRemoved);
```

#### getChunkHash

Returns the hash value of the chunk data.
```
function getChunkHash(bytes memory name, uint256 chunkId) external view returns (bytes32);
```

## Rationale

One issue of uploading the web contents to the blockchain is that the web contents may be too large to fit into a single transaction.  As a result, the standard provides a chunk-based operations so that uploading a content can be splitted into several transactions.  Meanwhile, the read operation can be done in a single transaction, i.e., with a single Web3 URL defined in EIP-4804.

## Security Considerations

No security considerations were found.

## Copyright

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 2: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE).
```

`read` method should return the concatenated chunked data written by `writeChunk` method.  The following gives some examples of the interactions:

- `read("hello.txt")` => "" (file is empty)
- `writeChunk("hello.txt", 0, "abc")`  will succeed
- `read("hello.txt")` => "abc"
- `writeChunk("hello.txt", 1, "efg")`  will succeed
- `read("hello.txt")` => "abcefg"
- `writeChunk("hello.txt", 0, "aaa")`  will succeed (replace chunk 0's data)
- `read("hello.txt")` => "aaaefg"
- `writeChunk("hello.txt", 3, "hij")`  will fail because the operation is not replacement or append.

With `writeChunk` method, we allow writing a file with external data that exceeds the current calldata limit (e.g., 1.8MB now, and 1.4MB post-EIP-4488), and it is able to read the whole file in a single `read` method (which is friendly for large web objects such as HTML/SVG/PNG/JPG, etc).

For `write` method, calling a `write` method will replace all data chunks of the file with `write` method data, and one implementation can be:
1. `writeChunk(filename, chunkId=0, data_from_write)` to chunk 0 with the same `write` method data; and
2. `truncate(filename, chunkId=1)`, which will remove the rest chunks.

> 

Reviewer 3: Missing required backward compatibility section.

Author: Thanks for the comment.  The backward compatibility section was added.

---
eip: 5018
title: Filesystem-like Interface for Smart Contracts
description: An interface to provide access to binary objects similar to filesystems.
author: Qi Zhou (@qizhou)
discussions-to: https://ethereum-magicians.org/t/eip-5018-directory-standard/8958
status: Draft
type: Standards Track
category: ERC
created: 2022-04-18
---

## Abstract

The following standardizes an API for directories and files within smart contracts, similar to traditional filesystems.
This standard provides basic functionality to read/write binary objects of any size, as well as allow reading/writing chunks of the object if the object is too large to fit in a single transaction.

## Motivation

A standard interface allows any binary objects on EVM-based blockchain to be re-used by other dApps.

> With [EIP-4804](./eip-4804.md), we are able to locate a Web3 resource on blockchain using HTTP-style URIs.  One application of Web3 resources are web contents that are referenced within a directory using relative paths such as HTML/SVG.  This standard proposes a contract-based directory to simplify the mapping between local web contents and on-chain web contents.  Further, with relative paths referenced in the web contents and EIP-4804, the users will have a consistent view of the web contents locally and on-chain.

Author: Done. Thanks for the comment.

```

#### getChunkHash

Returns the hash value of the chunk data.
```
function getChunkHash(bytes memory name, uint256 chunkId) external view returns (bytes32);
```

## Rationale

One issue of uploading the web contents to the blockchain is that the web contents may be too large to fit into a single transaction.  As a result, the standard provides chunk-based operations so that uploading a content can be splitted into several transactions.  Meanwhile, the read operation can be done in a single transaction, i.e., with a single Web3 URL defined in EIP-4804.

### Interactions Between Unchunked/Chunked Functions

`read` method should return the concatenated chunked data written by `writeChunk` method.  The following gives some examples of the interactions:

- `read("hello.txt")` => "" (file is empty)
- `writeChunk("hello.txt", 0, "abc")`  will succeed
- `read("hello.txt")` => "abc"
- `writeChunk("hello.txt", 1, "efg")`  will succeed
- `read("hello.txt")` => "abcefg"
- `writeChunk("hello.txt", 0, "aaa")`  will succeed (replace chunk 0's data)
- `read("hello.txt")` => "aaaefg"
- `writeChunk("hello.txt", 3, "hij")`  will fail because the operation is not replacement or append.

> With `writeChunk` method, we allow writing a file with external data that exceeds the current calldata limit (e.g., 1.8MB now, and 1.4MB post-EIP-4488), and it is able to read the whole file in a single `read` method (which is friendly for large web objects such as HTML/SVG/PNG/JPG, etc).

Reviewer 3: Since EIP-4488 is not final or expected to go into a fork, it should not be reference from another EIP.

Author: Done. EIP-4488 is removed. Thanks for the comment.
